sorting/bubble_sort_optimized.cpp

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file bubble_sort.cpp
+ * @brief Implementation of the Bubble Sort algorithm with self-tests.
+ *
+ * This file contains an implementation of the bubble sort algorithm
+ * inside the `sorting` namespace. It includes an optimized version
+ * of bubble sort that reduces the number of passes when the input
+ * is already sorted or becomes sorted before completing all iterations.
+ *
+ * Additionally, it contains a set of self-tests to verify correctness
+ * using the C++ standard library's `assert` mechanism.
+ *
+ * Example:
+ * @code
+ * std::vector<int> nums = {5, 4, 3, 2, 1};
+ * sorting::bubble_sort(nums);
+ * // nums becomes {1, 2, 3, 4, 5}
+ * @endcode
+ *
+ * @author Rohith A K
+ * @date 2025-09-25
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <utility>   /// for std::swap
+#include <vector>    /// for std::vector
+
+namespace sorting {
+
+/**
+ * @brief Sorts the given vector using the bubble sort algorithm.
+ *
+ * This optimized bubble sort reduces the number of passes in average cases
+ * by checking if the array is already sorted during each pass.
+ *
+ * @param nums the vector of integers to be sorted
+ * @return reference to the sorted vector
+ */
+std::vector<int>& bubble_sort(std::vector<int>& nums) {
```

Comment:
Could you try to make it generic by using templates to accommodate comparable data types?